File: agri_bot_perception/scripts/perception.py
#!/usr/bin/env python3

# Import the necessary libraries

import rospy  # Python library for ROS
from sensor_msgs.msg import Image  # Image is the message type
from cv_bridge import CvBridge  # Package to convert between ROS and OpenCV Images
import cv2  # OpenCV library
import numpy as np
import tf2_ros
import tf_conversions
import geometry_msgs
from std_msgs.msg import Int16

redLower = (0, 12, 12)
redUpper = (0, 213, 225)

cy = 240.5  # cy and cx are coordinates for center of camers
cx = 320.5
f = 554.387  # Focal length

img = None
dpt = None

# Colour codes for logging
GREEN = '\033[92m'
BOLD = '\033[1m'
UNDERLINE = '\033[4m'
END = '\033[0m'

br = CvBridge()

broadcaster = tf2_ros.TransformBroadcaster()


def img_callback(msg):  # RGB Image callback function
    global img
    img = br.imgmsg_to_cv2(msg, 'bgr8')
    img = np.array(img)


def depth_callback(dpt_msg):  # Depth Image callback function
    global dpt
    dpt = br.imgmsg_to_cv2(dpt_msg, '32FC1')
    dpt = np.array(dpt, dtype=np.dtype('float32'))


def main():

    count_feed = rospy.Publisher(
        'perception', Int16,
        queue_size=10)  # Publisher node to publish count of tomatoes

    rospy.Subscriber("/camera/color/image_raw2", Image,
                     img_callback)  # Subscribe to camera topics
    rospy.Subscriber("/camera/depth/image_raw2", Image, depth_callback)

    rate = rospy.Rate(10)  # 10Hz

    rospy.loginfo(
        f' {GREEN} {BOLD} {UNDERLINE}  Image Feed Recieved.... {END}')
    count_tom = Int16()

    while not rospy.is_shutdown():
        if (type(img) == np.ndarray) and (type(dpt) == np.ndarray):
            try:
                img2 = cv2.cvtColor(img, cv2.COLOR_BGR2HSV)

                mask = cv2.inRange(img2, redLower, redUpper)
                mask = cv2.erode(mask, None, iterations=1)
                mask = cv2.dilate(mask, None, iterations=6)

                contours, _ = cv2.findContours(mask, cv2.RETR_TREE,
                                               cv2.CHAIN_APPROX_NONE)
                count = 0
                contours = list(contours)

                for contour in contours:
                    (x, y), _ = cv2.minEnclosingCircle(contour)
                    x = int(x)
                    y = int(y)

                    text = 'Tomato'
                    font = cv2.FONT_HERSHEY_SIMPLEX
                    org = (x, y)
                    fontScale = 0.5
                    color = (255, 0, 0)
                    thickness = 1

                    depth = dpt[y, x]

                    # World Coordinates
                    X = depth * (x - cx) / f
                    Y = depth * (y - cy) / f
                    Z = depth

                    if Z < 1:  # Publish the transform only if the range is within the limit.
                        # Frame Id: camera_depth_frame2
                        t = geometry_msgs.msg.TransformStamped()
                        t.header.stamp = rospy.Time.now()
                        t.header.frame_id = 'camera_depth_frame2'
                        t.child_frame_id = "obj" + str(count)

                        # putting world coordinates coordinates as viewed for depth frame
                        t.transform.translation.x = X
                        t.transform.translation.y = Y
                        t.transform.translation.z = Z

                        # not extracting any orientation thus orientation is (0, 0, 0)
                        q = tf_conversions.transformations.quaternion_from_euler(
                            0, 0, 0)
                        t.transform.rotation.x = q[0]
                        t.transform.rotation.y = q[1]
                        t.transform.rotation.z = q[2]
                        t.transform.rotation.w = q[3]

                        if (X != 0) and (Y != 0) and (Z != 0):
                            broadcaster.sendTransform(t)
                        count += 1
                    cv2.putText(img, text, org, font, fontScale, color,
                                thickness,
                                cv2.LINE_AA)  # Put the text on image frame
                    cv2.drawContours(img, contour, -1, (0, 255, 0), 3)

                cv2.imshow("RGB IMAGE", img)

                cv2.waitKey(1)
                try:
                    if Z < 1:
                        count_tom.data = len(contours)
                        count_feed.publish(
                            count_tom
                        )  # Updates the manipulation node about number of tomatoes.

                except UnboundLocalError:
                    count_tom.data = 0
                    count_feed.publish(count_tom)

            except KeyboardInterrupt:
                print('Shutting Down')
                cv2.destroyAllWindows()
                break
            except:
                raise

        rate.sleep()


if __name__ == '__main__':
    rospy.init_node('perception',
                    anonymous=True)  # "perception" node initilization

    main()

File: agri_bot_perception/src/perception.cpp
#include <ros/ros.h>
#include <sensor_msgs/Image.h>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/aruco.hpp>
#include <tf2_ros/buffer.h>
#include <tf2_ros/transform_broadcaster.h>
#include <tf2_ros/static_transform_broadcaster.h>
#include <geometry_msgs/Pose.h>
#include <geometry_msgs/PoseStamped.h>
#include <std_msgs/Int16.h>
#include <std_msgs/Float64.h>
#include <iostream>
#include <boost/lexical_cast.hpp>
#include <std_msgs/String.h>



float cy = 240.5;
float cx = 320.5;
float f = 554.387;

float X;
float Y;
float Z;



class Perception{
    private:
        ros::NodeHandle node;
        ros::Subscriber image;
        ros::Subscriber depth_image;
        ros::Publisher count_feed;
        tf2_ros::TransformBroadcaster broadcaster;
        cv::InputArray redLower = (0,12,12);
        cv::InputArray redUpper = (0,213,225);
        cv::Mat dpt_img;
        

    public:
        Perception(){
            image = node.subscribe<sensor_msgs::Image>("/camera/color/image_raw2",1000,&Perception::imgCallback, this);
            depth_image = node.subscribe<sensor_msgs::Image>("/camera/depth/image_raw2",1000,&Perception::depthCallback, this);
            count_feed = node.advertise<std_msgs::Int16>("perception",1000,true);
        }

    void depthCallback(const sensor_msgs::ImageConstPtr& msg){
        cv_bridge::CvImagePtr dpt;
        dpt = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::TYPE_32FC1);
        dpt_img = dpt->image;
    }

    void imgCallback(const sensor_msgs::ImageConstPtr& msg){
        cv_bridge::CvImagePtr img;
        cv::Mat hsv_img;
        cv::Mat mask;
        std_msgs::Int16 count_tom; 
        // cv::OutputArray mask;
        
        img = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
        cv::cvtColor(img->image, hsv_img, cv::COLOR_BGR2HSV);
        std::vector<std::vector<cv::Point>> contours;
        cv::inRange(hsv_img, cv::Scalar(0,12,12), cv::Scalar(0,213,225), mask);

        cv::erode(mask, mask, cv::Mat(), cv::Point(-1,-1), 1);

        // cv::dilate(mask, mask, cv::Mat(), cv::Point(-1,-1), 6);

        cv::findContours(mask, contours,cv::RETR_TREE,cv::CHAIN_APPROX_NONE);
        int count = 0;

        for(int i = 0; i < contours.size(); i++){
            
            cv::Point2f center;
            float radius;
            cv::minEnclosingCircle(cv::InputArray(contours.at(i)), center, radius);
            
            std::string text = "Tomato";
            float fontScale = 0.5;
            cv::Scalar color = (255, 0, 0);
            int thickness = 1;
            float depth = dpt_img.at<float>(center.y, center.x);

            // std::cout << center.x << std::endl;
            // std::cout << center.y << std::endl;
            // std::cout << depth << std::endl;

            X = depth * (center.x - cx) / f;
            Y = depth * (center.y - cy) / f;
            Z = depth;
            std_msgs::String name;
            

            if ( Z < 0.9){
                name.data = "obj" + std::to_string(count);
                geometry_msgs::TransformStamped t;
                t.header.stamp = ros::Time::now();
                t.header.frame_id = "camera_depth_frame2";
                t.child_frame_id = name.data  ;

                t.transform.translation.x = X;
                t.transform.translation.y = Y;
                t.transform.translation.z = Z;

                t.transform.rotation.x = 0;
                t.transform.rotation.y = 0;
                t.transform.rotation.z = 0;
                t.transform.rotation.w = 1;

                broadcaster.sendTransform(t);
                count += 1;

            }
            cv::putText(img->image, text , center, cv::FONT_HERSHEY_SIMPLEX,fontScale, color, thickness, cv::LINE_AA);

        }
        cv::drawContours(img->image, contours, -1, (0,255,0), 3);
        cv::imshow("RGM Image", img->image);
        cv::waitKey(1);

        try{
            if (Z < 0.9){
                count_tom.data = contours.size();
                count_feed.publish(count_tom);
            }
        }
        catch(...){
            count_tom.data = 0;
            count_feed.publish(count_tom);          
        }
    }


};

int main(int argc, char** argv){

    ros::init(argc, argv, "Perception");

    Perception perception;

    ros::Rate rate(10);

    rate.sleep();
    ros::spin();
    

    return 0;
}

